java/arcs/tools/PlanGenerator.kt

```diff
@@ -0,0 +1,183 @@
+package arcs.tools
+
+import arcs.core.data.CountType
+import arcs.core.data.CreatableStorageKey
+import arcs.core.data.EntityType
+import arcs.core.data.FieldName
+import arcs.core.data.FieldType
+import arcs.core.data.Plan
+import arcs.core.data.Recipe
+import arcs.core.data.Schema
+import arcs.core.data.SchemaFields
+import arcs.core.data.SchemaName
+import arcs.core.data.TupleType
+import arcs.core.data.TypeVariable
+import arcs.core.storage.StorageKeyParser
+import arcs.core.type.Type
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.buildCodeBlock
+
+/**
+ * Translates a [Recipe] into code-generated [Plan] and [Plan.Handle] instances.
+ *
+ * @param builder A builder for output file
+ */
+fun Recipe.toGeneration(builder: FileSpec.Builder) {
```

Comment:
This whole comment is optional.. just thinking about API design here... but since it's an internal-only API it's not a big issue.

`toX` usually implies something is returned... Also, KotlinPoet/JavaPoet usage typically builds off of a builder. It would be slightly more idiomatic to flip this around to something like:

```kotlin
fun FileSpec.Builder.addRecipe(recipe: Recipe): FileSpec.Builder {
    // ...
}
```

Then the usage would look more fluent with the rest of the file's usage of the builder:

```kotlin
val fileHeader = FileSpec.builder(packageName, "")
	.addComment("""
        |GENERATED CODE -- DO NOT EDIT
        |
        |TODO(b/161941018): Improve whitespace / formatting.
    """.trimMargin())

val manifestProto = ManifestProto.parseFrom(manifest.readBytes())
val fileWithRecipes = manifestProto.decodeRecipes().filter { it.name != null }
    .fold(fileHeader) { builder, recipe -> builder.addRecipe(recipe) }
```

Similarly for the other methods in this file.


That said - I realize that's a large amount of rework to do, so I'm fine with things as-is if you want to keep making progress. Just a thought.

Reply:
I like it. It's a very open / close approach; it follows naturally from poet conventions. 

Happy to do the re-work so long as we get this right. 

@welnaseth What do you think about adopting this pattern for ArcsChron?

Reply:
Implemented.

Reply:
I'm definitely for more idiomatic usage patterns, I'll try to do this going forward and then I'll go back and fix the previous code to match.